File: LeetCode/Java Solutions/Most Stones Removed with Same Row or Column/Most Profit Assigning Work/stones.java
class Solution {
    class UnionFind {
        Map<Integer, Integer> parent = new HashMap<>();
        
        int find(int x) {
            if (!parent.containsKey(x)) {
                parent.put(x, x);
            } else if (parent.get(x) != x) {
                parent.put(x, find(parent.get(x)));
            }
            return parent.get(x);
        }
        
        void union(int x, int y) {
            int rootX = find(x);
            int rootY = find(y);
            if (rootX != rootY) {
                parent.put(rootX, rootY);
            }
        }
    }
    
    public int removeStones(int[][] stones) {
        UnionFind uf = new UnionFind();
        for (int[] stone : stones) {
            uf.union(stone[0], stone[1] + 10001);  // offset y to avoid collision with x
        }
        
        Set<Integer> groups = new HashSet<>();
        for (int[] stone : stones) {
            groups.add(uf.find(stone[0]));
            groups.add(uf.find(stone[1] + 10001));
        }
        
        return stones.length - groups.size();
    }
}

File: LeetCode/Python Solutions/Most Stones Removed with Same Row or Column/Most Profit Assigning Work/stones.py
class UnionFind:
    def __init__(self):
        self.parent = {}
        self.rank = {}

    def find(self, x):
        if x not in self.parent:
            self.parent[x] = x
            self.rank[x] = 0
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def union(self, x, y):
        px, py = self.find(x), self.find(y)
        if px == py:
            return
        if self.rank[px] < self.rank[py]:
            self.parent[px] = py
        elif self.rank[px] > self.rank[py]:
            self.parent[py] = px
        else:
            self.parent[py] = px
            self.rank[px] += 1

class Solution:
    def removeStones(self, stones: List[List[int]]) -> int:
        uf = UnionFind()
        for x, y in stones:
            uf.union(x, y + 10001)  # Offset y to avoid collision with x
        
        components = set(uf.find(x) for x, y in stones) | set(uf.find(y + 10001) for x, y in stones)
        return len(stones) - len(components)